src/module.c

```diff
@@ -7418,6 +7443,195 @@ RedisModuleBlockedClient *moduleBlockClient(RedisModuleCtx *ctx, RedisModuleCmdF
     return bc;
 }
 
+/* This API registers a callback to execute in addition to normal password based authentication.
+ * Multiple callbacks can be registered across different modules. When a Module is unloaded, all the
+ * auth callbacks registered by it are unregistered.
+ * The callbacks are attempted (in the order of most recently registered first) when the AUTH/HELLO
+ * (with AUTH field is provided) commands are called.
+ * The callbacks will be called with a module context along with a username and a password, and are
+ * expected to take one of the following actions:
+ * (1) Authenticate - Use the RM_AuthenticateClient* API and return REDISMODULE_AUTH_HANDLED.
+ * This will immediately end the auth chain as successful and add the OK reply.
+ * (2) Deny Authentication - Return REDISMODULE_AUTH_HANDLED without authenticating or blocking the
+ * client. Optionally, `err` can be set to a custom error message and `err` will be automatically
+ * freed by the server.
+ * This will immediately end the auth chain as unsuccessful and add the ERR reply.
+ * (3) Block a client on authentication - Use the RM_BlockClientOnAuth API and return
+ * REDISMODULE_AUTH_HANDLED. Here, the client will be blocked until the RM_UnblockClient API is used
+ * which will trigger the auth reply callback (provided through the RM_BlockClientOnAuth).
+ * In this reply callback, the Module should authenticate, deny or skip handling authentication.
+ * (4) Skip handling Authentication - Return REDISMODULE_AUTH_NOT_HANDLED without blocking the
+ * client. This will allow the engine to attempt the next custom auth callback.
+ * If none of the callbacks authenticate or deny auth, then password based auth is attempted and
+ * will authenticate or add failure logs and reply to the clients accordingly.
+ *
+ * Note: If a client is disconnected while it was in the middle of blocking custom auth, that
+ * occurrence of the AUTH or HELLO command will not be tracked in the INFO command stats.
+ *
+ * The following is an example of how non-blocking module based authentication can be used:
+ *
+ *      int auth_cb(RedisModuleCtx *ctx, RedisModuleString *username, RedisModuleString *password, RedisModuleString **err) {
+ *           const char* user = RedisModule_StringPtrLen(username, NULL);
+ *           const char* pwd = RedisModule_StringPtrLen(password, NULL);
+ *           if (!strcmp(user,"foo") && !strcmp(pwd,"valid_password")) {
+ *               RedisModule_AuthenticateClientWithACLUser(ctx, "foo", 3, NULL, NULL, NULL);
+ *               return REDISMODULE_AUTH_HANDLED;
+ *           }
+ *           else if (!strcmp(user,"foo") && !strcmp(pwd,"wrong_password")) {
+ *               RedisModuleUser *user = RedisModule_GetModuleUserFromUserName(username);
+ *               if (user) {
```

Comment:
This is clunky, and kind of awkward since end modules might actually want to log failures for users that don't exist. I'm not sure if we want to add a new API just for ACL errors, that doesn't actually require the user to exist, similar to how we do it internally. Thoughts?

Reply:
Yes. Right now, during custom auth, if a Module checks the credentials and finds them to be invalid, it still has to create a RedisModuleUser just so that it can use the `RedisModule_ACLAddLogEntry ` API to add a ACL error log entry.

This API requires the RedisModuleUser to be provided, but it only uses this to obtain the username. Instead it could accept the username as a RedisModuleString.
```
int RM_ACLAddLogEntry(RedisModuleCtx *ctx, RedisModuleUser *user, RedisModuleString *object, RedisModuleACLLogEntryReason reason) {
    int acl_reason;
    switch (reason) {
        case REDISMODULE_ACL_LOG_AUTH: acl_reason = ACL_DENIED_AUTH; break;
        case REDISMODULE_ACL_LOG_KEY: acl_reason = ACL_DENIED_KEY; break;
        case REDISMODULE_ACL_LOG_CHANNEL: acl_reason = ACL_DENIED_CHANNEL; break;
        case REDISMODULE_ACL_LOG_CMD: acl_reason = ACL_DENIED_CMD; break;
        default: return REDISMODULE_ERR;
    }

    addACLLogEntry(ctx->client, acl_reason, ACL_LOG_CTX_MODULE, -1, user->user->name, sdsdup(object->ptr));
    return REDISMODULE_OK;
}
```

But changing this API will be another breaking change

Reply:
we can't break the API, but i suppose we need to add another one.
if we do that, we can consider allowing a NULL `object`.

Reply:
I'm inclined to just leave it as is, it's a bit clunky, but it's not really that much extra inefficiency.

Reply:
Added the new API here: https://github.com/redis/redis/pull/11659/commits/4a6c0b0344cea370b3b468a708975162efaaead6